### PR TITLE
feat: Support OAuth for all plugin types

### LINKED ISSUE #7973: [Feature] Add support for OAuth2 type datasource forms on client side
## Summary

All datasource forms today are rendered by config. This works for most authentication types, but since OAuth2 needs user interaction across multiple endpoints, we cannot render this form from config.

This feature request is to create a generic type that handles authentication for OAuth2. This should be an extensible solution that works in the future for other authentication mechanisms like OIDC.

### PR DESCRIPTION
**Pull request by ohansFavour**
## Description

In this feature, 

- The authentication workflow for OAuth2 has been made generic
- The call-to-actions available in data source forms are now fully customizable depending on the authentication type of the plugin. 

In addition to these, the PR fixes the following

- When generating a new page from Gsheet, I don't get redirected to the "generate page from data" page.
- Route changes before Datasource deletion is successful. Sometimes, I am routed to the integrations page even when the Gsheet Datasource deletion wasn't successful. 
- Routing to a deleted Gsheet data source editor page shows bizarre UI.

Fixes #7973

## Type of change

- New feature (non-breaking change which adds functionality)

## How Has This Been Tested?

Manual

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

## Test coverage results :test_tube:
<details><summary>:red_circle: Total coverage has decreased</summary>

    // Code coverage diff between base branch:release and head branch: feat/support-oauth-for-all-plugins 
Status | File | % Stmts | % Branch | % Funcs | % Lines 
 -----|-----|---------|----------|---------|------ 
 :red_circle: | total | 54.99 **(-0.02)** | 36.69 **(-0.02)** | 34.71 **(0.01)** | 55.51 **(0)**
 :sparkles: :new: | **app/client/src/api/OAuthApi.ts** | **80** | **100** | **33.33** | **77.78**
 :green_circle: | app/client/src/entities/Datasource/index.ts | 52.17 **(20.92)** | 33.33 **(33.33)** | 50 **(50)** | 58.82 **(28.82)**
 :green_circle: | app/client/src/pages/Editor/DataSourceEditor/DBForm.tsx | 54.84 **(7.47)** | 22.22 **(-7.41)** | 9.09 **(1.4)** | 60 **(12.7)**
 :red_circle: | app/client/src/pages/Editor/SaaSEditor/DatasourceForm.tsx | 38.71 **(-0.89)** | 11.76 **(-7.75)** | 14.29 **(3.18)** | 43.4 **(1.91)**
 :sparkles: :new: | **app/client/src/pages/common/datasourceAuth/DefaultAuth.tsx** | **42.11** | **33.33** | **0** | **44.44**
 :sparkles: :new: | **app/client/src/pages/common/datasourceAuth/OAuth.tsx** | **39.62** | **23.08** | **14.29** | **40.38**
 :sparkles: :new: | **app/client/src/pages/common/datasourceAuth/index.tsx** | **33.33** | **0** | **0** | **33.33**
 :red_circle: | app/client/src/sagas/DatasourcesSagas.ts | 13.29 **(-0.12)** | 2.4 **(0)** | 8.7 **(-0.39)** | 15.75 **(-0.12)**
 :red_circle: | app/client/src/selectors/entitiesSelector.ts | 52.68 **(-0.84)** | 12.38 **(-0.75)** | 32.31 **(-0.76)** | 49.8 **(-1.01)**
 :green_circle: | app/client/src/utils/WorkerUtil.ts | 89.76 **(0.78)** | 72.55 **(1.96)** | 100 **(0)** | 93.33 **(0.95)**
 :x: | ~~app/client/src/api/SaasApi.ts~~ | ~~80~~ | ~~100~~ | ~~33.33~~ | ~~77.78~~</details>

### CODE REVIEW

File: app/client/src/pages/Editor/DataSourceEditor/DBForm.tsx [hunk 1]
```diff
@@ -110,34 +88,17 @@ class DatasourceDBEditor extends JSONtoForm<Props> {
...
-    const {
-      datasourceId,
-      handleDelete,
-      isDeleting,
-      isSaving,
-      isTesting,
-      messages,
-      pluginType,
-    } = this.props;
+    const { datasource, formData, messages, pluginType } = this.props;
```
Comment (Irongade) [inline app/client/src/pages/Editor/DataSourceEditor/DBForm.tsx:99]: Can we add viewMode prop to this line as well?

**Comment by Irongade** [comment]
can we add a short description of what this feature is about for more context?

**Review by ayushpahwa — CHANGES_REQUESTED** [review]
The code looks good. Please add some comments here for ease of understanding in the future.

**Review by Irongade — APPROVED** [review]

**Review by Irongade — APPROVED** [review]

**Comment by ohansFavour** [comment]
> The code looks good. Please add some comments here for ease of understanding in the future.

Done

**Review by ayushpahwa — APPROVED** [review]

**Pull request merged**